File: admin/frontend/src/api/role.ts
import type { PageData } from '@/utils/request';
import { del, download, get, post, put } from '@/utils/request';

export interface Role {
  id: number;
  code: string;
  name: string;
  description: string;
  sort: number;
  status: 1 | 2;
  created_at: string;
  updated_at: string;
  menu_ids: number[];
}

export interface CreateRoleReq {
  code: string;
  name: string;
  description?: string;
  sort?: number;
  status?: 1 | 2;
  menu_ids?: number[];
}

function qs(params: Record<string, unknown>): string {
  const usp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null && v !== '') usp.set(k, String(v));
  }
  return usp.toString();
}

export const roleApi = {
  list: (page: number, size: number, keyword?: string) =>
    get<PageData<Role>>(`/system/roles?${qs({ page, size, keyword })}`),
  get: (id: number) => get<Role>(`/system/roles/${id}`),
  create: (body: CreateRoleReq) => post<Role>('/system/roles', body),
  update: (id: number, body: Partial<CreateRoleReq>) => put<Role>(`/system/roles/${id}`, body),
  remove: (id: number) => del(`/system/roles/${id}`),
  batchRemove: (ids: number[]) => post<null>('/system/roles/batch-delete', { ids }),
  assignMenus: (id: number, menuIds: number[]) =>
    post<null>(`/system/roles/${id}/menus`, { menu_ids: menuIds }),
  exportExcel: (keyword?: string) => download(`/system/roles/export/excel?${qs({ keyword })}`),
  exportCSV: (keyword?: string) => download(`/system/roles/export/csv?${qs({ keyword })}`),
};

File: admin/backend/internal/dto/role.go
package dto

// CreateRoleReq is the body of POST /system/roles.
type CreateRoleReq struct {
	Code        string   `json:"code"        binding:"required,min=1,max=64"`
	Name        string   `json:"name"        binding:"required,min=1,max=64"`
	Description string   `json:"description" binding:"max=255"`
	Sort        int      `json:"sort"`
	Status      int8     `json:"status"      binding:"omitempty,oneof=1 2"`
	MenuIDs     []uint64 `json:"menu_ids"`
}

// UpdateRoleReq is the body of PUT /system/roles/:id.
type UpdateRoleReq struct {
	Name        *string  `json:"name"        binding:"omitempty,max=64"`
	Description *string  `json:"description" binding:"max=255"`
	Sort        *int     `json:"sort"`
	Status      *int8    `json:"status"      binding:"omitempty,oneof=1 2"`
	MenuIDs     []uint64 `json:"menu_ids"`
}

// RoleResp is the wire shape for a single role.
type RoleResp struct {
	ID          uint64   `json:"id"`
	Code        string   `json:"code"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Sort        int      `json:"sort"`
	Status      int8     `json:"status"`
	CreatedAt   string   `json:"created_at"`
	UpdatedAt   string   `json:"updated_at"`
	MenuIDs     []uint64 `json:"menu_ids"`
}

// AssignMenusReq is the body of POST /system/roles/:id/menus.
type AssignMenusReq struct {
	MenuIDs []uint64 `json:"menu_ids" binding:"required"`
}

// AssignUsersReq is the body of POST /system/roles/:id/users.
type AssignUsersReq struct {
	UserIDs []uint64 `json:"user_ids" binding:"required"`
}
